geniza/corpus/management/commands/sync_annotation_export.py

```diff
@@ -1,22 +1,132 @@
+import json
+import os.path
+from collections import defaultdict
+from datetime import datetime
+
 from django.conf import settings
+from django.contrib.admin.models import LogEntry
+from django.contrib.contenttypes.models import ContentType
 from django.core.management.base import BaseCommand
+from django.utils import timezone
 
+from geniza.annotations.models import Annotation
 from geniza.corpus.annotation_export import AnnotationExporter
+from geniza.corpus.models import Document
 
 
 class Command(BaseCommand):
     """Synchronize annotation backup data with GitHub"""
 
+    # filename for last run information (stored in current user's home)
+    lastrun_filename = os.path.join(os.path.expanduser("~"), ".pgp_export_lastrun")
+    # id for this script in the last run file info
+    script_id = "annotations"
+
+    #: default verbosity
+    v_normal = 1
+
     def handle(self, *args, **options):
         if not getattr(settings, "ANNOTATION_BACKUP_PATH"):
             raise CommandError(
                 "Please configure ANNOTATION_BACKUP_PATH in django settings"
             )
 
-        anno_exporter = AnnotationExporter(
-            stdout=self.stdout, verbosity=options["verbosity"]
+        # initialize annotation exporter; don't push changes automatically
+        self.anno_exporter = AnnotationExporter(
+            stdout=self.stdout, verbosity=options["verbosity"], push_changes=False
         )
-        # set up repo object (pulls any changes)
-        anno_exporter.setup_repo()
-        # push changes
-        anno_exporter.sync_github()
+        # set up repo object and pull any changes
+        self.anno_exporter.setup_repo()
+
+        # determine last run
+        lastrun = self.script_lastrun()
+        # get annotation log entries since the last run
+        annotation_ctype = ContentType.objects.get_for_model(Annotation)
+        # get all log entries for changes on annotations since the last run
+        log_entries = LogEntry.objects.filter(
+            content_type_id=annotation_ctype.pk, action_time__gte=lastrun
+        )
+        # store the datetime immediately after this query for the next run
+        new_lastrun = timezone.now()
+
+        if options["verbosity"] >= self.v_normal:
+            print("%d annotation log entries since %s" % (log_entries.count(), lastrun))
+
+        # generate exports based on what has been changed
+        if log_entries.exists():
+
+            # TODO: handle deletions (object no longer in db)
+
+            # make a dictionary of modified annotations and the users who modified them
+            modified_annotations = defaultdict(list)
+            for log_entry in log_entries:
```

Comment:
This block could maybe be a method, though I could also see wanting to reduce the amount of stuff needed to pass around, and wanting to call `self.anno_exporter.export` in this main handle method. But it could make testing easier maybe.